### PR TITLE
Alugrid2

### PR DESCRIPTION
**Pull request by ElyesAhmed**
This is an attempt to integrate Dune ALUGrid with opm. In this first PR, we disabled Space_Filling Curve_Ordering. For ease of merging we leave these features ( incomplete yet) for the next PR.

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
I'm happy to see that this PR compiles on the CI system where there is no ALUGrid.  There are, however, a number of [new diagnostics](https://ci.opm-project.org/job/opm-simulators-PR-builder/3802/gcc/new/) being emitted by the compiler and we should address those before we merge.  I'll do a more thorough review later.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
Thanks a lot for addressing the warnings.  I'll perform a real review later today CEST.

**Comment by bska** [comment]
jenkins build this please

**Review by bska — COMMENTED** [review]
I think is a decent first step, but the PR appears to completely disable the [`PolyhedralGrid`](https://github.com/OPM/opm-grid/blob/98ce09b84d79d13d3b0ac503ef02e039cde5e4df/opm/grid/polyhedralgrid/grid.hh#L153) support in OPM Flow.  Is that true?  If so, I'd like to know if there's a plan to reenable this feature at some point.

There are also a number of other, smaller, issues that I'd like to know more about before merging this.

**Review by blattms — COMMENTED** [review]
I might be missing something of course, but:
the aim of introducing the `CartesianIndexMappers` was to get rid of using `globalCell` arrays of `CpGrid`. This PR now introduces these in the vanguards. Is that really needed or can we make this work also by just using the `CartesianIndexMappers`?

**Comment by ElyesAhmed** [comment]
> 

Thank you Markus for your feedback, yes I agree that we should use CartesianIndexMapper however as we are not allowing re-ordering yet in this first PR, and we will change that when we add space filling curve ordering (SFR) in the next step.

File: ebos/eclalugridvanguard.hh [hunk 1]
```diff
@@ -200,42 +296,74 @@ protected:
...
+          input_grid = &this->eclState().getInputGrid();
+          global_porv = this->eclState().fieldProps().porv(true);
+          OpmLog::info("\nProcessing grid");
+          }
+
+#if HAVE_MPI
+        this->equilGrid_ = std::make_unique<Dune::CpGrid>(EclGenericVanguard::comm());
+#else
+        this->equilGrid_ = std::make_unique<Dune::CpGrid>();
+#endif
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply ebos/eclalugridvanguard.hh:315]: yes no need to that check

File: ebos/eclbasevanguard.hh [hunk 2]
```diff
@@ -503,14 +507,14 @@ protected:
         const auto num_aqu_cells = this->allAquiferCells();
 
         for (; elemIt != elemEndIt; ++elemIt) {
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply ebos/eclbasevanguard.hh:509]: done

File: ebos/eclproblem.hh [hunk 3]
```diff
@@ -867,7 +862,23 @@ public:
...
+        std::vector<unsigned int> GridToequilGrid;
+        #ifdef HAVE_DUNE_ALUGRID
+        using Grid = GetPropType<TypeTag, Properties::Grid>;
+        typename std::is_same<Grid, Dune::ALUGrid<3, 3, Dune::cube, Dune::nonconforming>>::type isAlugrid;
+
+        if (isAlugrid) {
+          for (size_t index = 0; index<simulator.vanguard().cartesianSize(); index ++) {
+               equilGridToGrid.push_back(simulator.vanguard().gridEquilIdxToGridIdx(index));
+               GridToequilGrid.push_back(simulator.vanguard().gridIdxToEquilGridIdx(index));
+           }
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply ebos/eclproblem.hh:877]: fixed

File: ebos/eclproblem.hh [hunk 4]
```diff
@@ -1161,13 +1184,27 @@ public:
         auto& schedule = simulator.vanguard().schedule();
         auto& ecl_state = simulator.vanguard().eclState();
         int episodeIdx = this->episodeIndex();
+        // re-ordering in case of ALugrid
+        std::vector<unsigned int> GridToequilGrid;
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply ebos/eclproblem.hh:1188]: changed now everywhere

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 5]
```diff
@@ -1627,8 +1631,20 @@ namespace Opm {
...
+ //   {
+ //       const auto& cartMapper = ebosSimulator_.vanguard().cartesianIndexMapper();
+ //       const size_t cartesianSize = cartMapper.cartesianSize();
+ //       cartesian_to_compressed_.resize(cartesianSize, -1);
+
+//        for (unsigned i = 0; i < local_num_cells_; ++i) {
+//            unsigned cartesianCellIdx = cartMapper.cartesianIndex(i);
+//            cartesian_to_compressed_[cartesianCellIdx] = i;
+//        }
+ //   }
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:1647]: cleaned, thanks!

File: CMakeLists.txt [hunk 6]
```diff
@@ -489,6 +490,23 @@ opm_add_test(flow_distribute_z
   ${FLOW_TGTS}
   $<TARGET_OBJECTS:moduleVersion>
   )
+if (NOT BUILD_FLOW_ALU_GRID)
+  set(FLOW_ALUGRID_ONLY_DEFAULT_ENABLE_IF "FALSE")
+else()
+  set(FLOW_ALUGRID_ONLY_DEFAULT_ENABLE_IF "TRUE")
+endif()
+
+# the production oriented general-purpose ECL simulator
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply CMakeLists.txt:499]: fixed now

File: opm/simulators/flow/Main.hpp [hunk 7]
```diff
@@ -145,6 +145,7 @@ class Main
         initMPI();
     }
 
+#define DEMONSTRATE_RUN_WITH_NONWORLD_COMM 1
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply opm/simulators/flow/Main.hpp:148]: thanks, it was changed while modifying another branch!

File: ebos/eclgenericwriter.cc [hunk 8]
```diff
@@ -578,18 +592,59 @@ template class EclGenericWriter<Dune::CpGrid,
...
                                 double>;
 
+#endif // HAVE_DUNE_ALUGRID
+#endif // !HAVE_DUNE_FEM
+
+//template class EclGenericWriter<Dune::PolyhedralGrid<3,3,double>,
+//                                Dune::PolyhedralGrid<3,3,double>,
+//                                Dune::GridView<Dune::PolyhedralGridViewTraits<3, 3, double, Dune::PartitionIteratorType(4)>>,                              Dune::MultipleCodimMultipleGeomTypeMapper<Dune::GridView<Dune::PolyhedralGridViewTraits<3,3,double,Dune::PartitionIteratorType(4)>>>,
+//                                double>;
+
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply ebos/eclgenericwriter.cc:649]: fixed

File: options.cmake [hunk 9]
```diff
@@ -0,0 +1,4 @@
+set(CMAKE_BUILD_TYPE "Debug" CACHE STRING "Build type")
+set(USE_OPENMP OFF CACHE BOOL "Do not use OpenMP")
+set(USE_MPI OFF CACHE BOOL "Do not use MPI")
+set(USE_SUPERLU OFF CACHE BOOL "Do not use SUPERLU")
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply options.cmake:4]: yes my private options! I will delete this

File: ebos/eclalugridvanguard.hh [hunk 10]
```diff
@@ -200,42 +296,74 @@ protected:
...
+        factory_ = std::make_unique<Factory>();
+        grid_ = factory_->convert(*equilGrid_, cartesianCellId_, ordering_);
+        std::cerr << "WARNING: Space Filling Curve Ordering is not yet supported: DISABLE_ALUGRID_SFC_ORDERING is enabled" <<
+             std::endl;
+        equilGridToGrid_.resize(ordering_.size());
+        for (size_t index = 0; index<ordering_.size(); ++index) {
+            equilGridToGrid_[ordering_[index]] = index;
+        }
+
+        cartesianIndexMapper_.reset( new CartesianIndexMapper(*grid_, cartesianDimension_, cartesianCellId_));
```
[the remark this answers is not in the document]
Comment (ElyesAhmed) [reply ebos/eclalugridvanguard.hh:344]: fixed

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by blattms** [comment]
@ElyesAhmed I don't understand the commit message of eba85ac. there are a lot of comments there, but they seem unrelated to the changes. What is it about?

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
Merging after discussion with @bska and @ElyesAhmed.

**Pull request merged**